File: rainbow-lib/src/main/java/com/juankysoriano/rainbow/utils/schedulers/RainbowSchedulers.java
package com.juankysoriano.rainbow.utils.schedulers;

import java.util.concurrent.Executors;
import java.util.concurrent.ThreadFactory;

public abstract class RainbowSchedulers {
    public static RainbowScheduler single(String name, Priority priority) {
        ThreadFactory threadFactory = RainbowThreadFactory.newInstance(name, priority.threadPriority);
        return new RainbowScheduler(Executors.newScheduledThreadPool(1, threadFactory));
    }

    /**
     * The stack size for the threads on this scheduler is enough to perform long recursion tasks
     **/
    public static RainbowScheduler singleForRecursion(String name, Priority priority) {
        ThreadFactory threadFactory = RainbowThreadFactory.newInstanceForRecursion(name, priority.threadPriority);
        return new RainbowScheduler(Executors.newScheduledThreadPool(1, threadFactory));
    }

    public static RainbowScheduler multiThreaded(String name, Priority priority) {
        ThreadFactory threadFactory = RainbowThreadFactory.newInstance(name, priority.threadPriority);
        return new RainbowScheduler(Executors.newScheduledThreadPool(Runtime.getRuntime().availableProcessors(), threadFactory));
    }

    /**
     * The stack size for the threads on this scheduler is enough to perform long recursion tasks
     **/
    public static RainbowScheduler multiThreadedForRecursion(String name, Priority priority) {
        ThreadFactory threadFactory = RainbowThreadFactory.newInstanceForRecursion(name, priority.threadPriority);
        return new RainbowScheduler(Executors.newScheduledThreadPool(Runtime.getRuntime().availableProcessors(), threadFactory));
    }

    public enum Priority {
        MAX(Thread.MAX_PRIORITY),
        NORMAL(Thread.NORM_PRIORITY),
        MIN(Thread.MIN_PRIORITY);

        private int threadPriority;

        Priority(int threadPriority) {
            this.threadPriority = threadPriority;
        }
    }
}

File: rainbow-lib/src/main/java/com/juankysoriano/rainbow/core/schedulers/RainbowSchedulers.kt
package com.juankysoriano.rainbow.core.schedulers

import java.util.concurrent.Executors

object RainbowSchedulers {

    fun single(name: String, priority: Priority): RainbowScheduler {
        val threadFactory = RainbowThreadFactory.create(name, priority.threadPriority)
        return RainbowScheduler(Executors.newScheduledThreadPool(1, threadFactory))
    }

    /**
     * The stack size for the threads on this scheduler is enough to perform long recursion tasks
     */
    fun singleForRecursion(name: String, priority: Priority): RainbowScheduler {
        val threadFactory = RainbowThreadFactory.createForRecursion(name, priority.threadPriority)
        return RainbowScheduler(Executors.newScheduledThreadPool(1, threadFactory))
    }

    fun multiThreaded(name: String, priority: Priority): RainbowScheduler {
        val threadFactory = RainbowThreadFactory.create(name, priority.threadPriority)
        return RainbowScheduler(Executors.newScheduledThreadPool(Runtime.getRuntime().availableProcessors(), threadFactory))
    }

    /**
     * The stack size for the threads on this scheduler is enough to perform long recursion tasks
     */
    fun multiThreadedForRecursion(name: String, priority: Priority): RainbowScheduler {
        val threadFactory = RainbowThreadFactory.createForRecursion(name, priority.threadPriority)
        return RainbowScheduler(Executors.newScheduledThreadPool(Runtime.getRuntime().availableProcessors(), threadFactory))
    }

    enum class Priority private constructor(val threadPriority: Int) {
        MAX(Thread.MAX_PRIORITY),
        NORMAL(Thread.NORM_PRIORITY),
        MIN(Thread.MIN_PRIORITY)
    }
}
